### ISSUE TITLE
Unable to initialize BasicRegistrationFactory with external empty dictionary

### ISSUE THREAD

**Issue by m-charlton** [sunpy/sunpy#1988]
Not sure if this is a bug or a subtle implementation detail. 

Found this while writing unit tests for `sunpy.util.datatype_factory_base.BasicRegistrationFactory`.

The constructor of `BasicRegistrationFactory` allows us to pass an external
dictionary to be used as a registry.

In `sunpy/util/datatype_factory_base.py`:

```
 5  class BasicRegistrationFactory(object):
...
43  def __init__(self, default_widget_type=None,
44               additional_validation_functions=[], registry=None):
45
46      if registry:
47          self.registry = registry
48      else:
49          self.registry = dict()
```

If you pass an empty dictionary:

```
from collections import OrderedDict

external_registry = OrderedDict()

reg_factory = BasicRegistrationFactory(registry=external_registry)

```

Then line `47` in `BasicRegistrationFactory` is *not* hit and we end up with the 
registry being initialized at line `49`. If `external_registry` is not empty 
then the registry is initialized at line `47`, as expected.

If it is a bug then I'll submit a fix along with some updated unit tests.

**Comment by Cadair** [comment]
yeah that's a bug. Nice catch.

**Comment by nabobalis** [comment]
Closed by #1988
